### PR TITLE
Base support for Merkle proofs

### PR DESCRIPTION
**Pull request by td202**
## Purpose

This PR introduces a Haskell type for Merkle proofs, as well as parsers for interpreting such proofs.

It also introduces `BlockHashVersion` (cf. #490) and a new `TransactionOutcomesVersion`. Moreover, code related to transaction outcomes is relocated to `concordium-node`.

## Changes

- Add `Concordium.MerkleProof`:
  - `MerkleProof`: type of Merkle proofs.
  - `MerkleSchema`: type of schemas for parsing Merkle proofs.
  - `PartialTree`: parsed representation of a Merkle proof as a tagged tree of data.
  - `parseMerkleProof`: function for parsing a Merkle proof into a 'PartialTree' according to a schema.
  - `MerkleProvable`: typeclass for data that can produce a Merkle proof.
- Introduce `BlockHashVersion` (duplicates #490).
- Introduce `TOV2` as a new `TransactionOutcomesVersion` where the outcomes are hashed according to the new LFMBTree hashing scheme.
- Move definitions related to transaction outcomes into `concordium-node`.

## Checklist

- [X] My code follows the style of this project.
- [X] The code compiles without warnings.
- [X] I have performed a self-review of the changes.
- [X] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by EmilLa1 — COMMENTED** [review]
Some comments for now. 

I think it would be very valuable with some tests for this even though if the schema might be up for changes.

**Review by EmilLa1 — COMMENTED** [review]
A few more comments. Looking good!

File: haskell-src/Concordium/MerkleProofs.hs [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+      RawData !BS.ByteString
+    | -- | A subproof that is reduced to a hash before concatenating when computing the root hash.
+      SubProof !MerkleProof
+    deriving (Show)
+
+-- | Determine if a Merkle proof is empty (it contains no bytes).
+isEmpty :: MerkleProof -> Bool
+isEmpty [] = True
+isEmpty (RawData bytes : rest) = BS.null bytes && isEmpty rest
+isEmpty _ = False
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-src/Concordium/MerkleProofs.hs:57]: A `SubProof` always reduces to a hash, so is not an empty `MerkleProof`.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+
+-- | Compute the root hash represented by a Merkle proof.
+toRootHash :: MerkleProof -> SHA256.Hash
+toRootHash l = SHA256.hashLazy $ Builder.toLazyByteString $ mconcat $ merkleBuilder <$> l
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-src/Concordium/MerkleProofs.hs:61]: In principle, it avoids constructing an intermediate strict `ByteString` containing all the data to be hashed, so we avoid some copying.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+
+-- | A helper function for constructing a 'Builder.Builder' from a 'MerkleBranch'.
+-- This is used for managing branches for the purposes of computing root hashes.
+merkleBuilder :: MerkleBranch -> Builder.Builder
+merkleBuilder (RawData bytes) = Builder.byteString bytes
+merkleBuilder (SubProof sp) = Builder.shortByteString (SHA256.hashToShortByteString $ toRootHash sp)
+
+-- * Parsing Merkle proofs
+
+-- | A schematic type that describes how some data item is represented in a Merkle proof.
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-src/Concordium/MerkleProofs.hs:71]: It is a schema: a `MerkleData` value describes how data is represented in a Merkle tree, as distinct from being data from a Merkle tree.

**Review by limemloh — APPROVED** [review]
LGTM, only minor comments

File: haskell-src/Concordium/MerkleProofs.hs [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+        EQ -> cont bytes rest ctx
+        GT -> let (b0, b1) = BS.splitAt len bytes in cont b0 (RawData b1 : rest) ctx
+    inner parserError _ _ (SubProof _ : _) _ = parserError UnexpectedSubProof
+
+-- | Given a parser for a sub-proof, parse that sub-proof as part of a larger proof.
+--  The sub-proof parser should return (when successful) a builder for the byte string that should
+--  be hashed to give the root hash of the sub-proof.
+--  If a sub-proof is not next in the input, then the parser will take the next 32 bytes as a hash
+--  instead, treating the sub-proof as absent. If neither sub-proof nor 32 bytes are available, this
+--  will raise a 'ParseError'.
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:281]: I don't understand why this is reasonable behaviour. So it should be explained.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+                                then
+                                    let hsh = hashBuilder builder
+                                    in  parserCont
+                                            (hsh, Just subPT)
+                                            rest
+                                            parserReaderContext
+                                else parserError (applyErrorContext ctx ExpectedEndOfInput)
+                        }
+
+parseSize ::
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:317]: Documentation is missing.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+    (Num b) =>
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:318]: This is kind of too general. If you are parsing a U64 but give Word8 here the meaning here is not really as intended.

At least the preconditions must be documented.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+
+-- | Compute a hash from a 'Builder.Builder'.
+hashBuilder :: Builder.Builder -> SHA256.Hash
+hashBuilder = SHA256.hashLazy . Builder.toLazyByteString
+
+-- | Construct a 'Builder.Builder' from a 'SHA256.Hash'.
+builderHash :: SHA256.Hash -> Builder.Builder
+builderHash = Builder.shortByteString . SHA256.hashToShortByteString
+
+-- | Convert bytes into a number, parsing it big endian.
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:402]: Some information about overflow handling should be stated here.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 8]
```diff
@@ -0,0 +1,566 @@
...
+            setFresh . Sequence $
+                [ Hashed "transactionsAndOutcomes" transactionsAndOutcomesHash,
+                  Tagged "state" opaqueHash
+                ]
+        blockQuasiHash <- setFresh . Sequence $ [Hashed "meta" metaHash, Hashed "data" dataHash]
+        schemaAt blockHash ?= Sequence [Hashed "header" blockHeaderHash, Hashed "quasi" blockQuasiHash]
+        return blockHash
+
+-- | A class for types that can produce 'MerkleProof's.
+class MerkleProvable m t where
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:563]: Again, please document what the type parameters are meant to be, ideally naming them more meaningfully.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+                            Sequence
+                                [ Tagged "round" u64,
+                                  Tagged "finalizers" (VariableLengthBytes SizeU32BE)
+                                ],
+                          RepeatedBE "finalizerQCRoundsSecondEpoch" SizeU32BE $
+                            Sequence
+                                [ Tagged "round" u64,
+                                  Tagged "finalizers" (VariableLengthBytes SizeU32BE)
+                                ],
+                          Tagged "aggregateSignature" (FixedLengthBytes 48)
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:514]: Isn't there a constant for this defined in aggregate sig module?

File: haskell-src/Concordium/MerkleProofs.hs [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+                        [ LiteralBytes (S.encode (0 :: Word8)),
+                          Tagged "null" (FixedLengthBytes 0)
+                        ]
+                    )
+                    ( Sequence
+                        [ LiteralBytes (S.encode (1 :: Word8)),
+                          Hashed "finalizedBlock" blockHash,
+                          Tagged "finalizedRound" u64,
+                          Tagged "epoch" u64,
+                          Tagged "finalizedAggregateSignature" (FixedLengthBytes 48),
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:530]: Also here.

File: haskell-src/Concordium/MerkleProofs.hs [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+      VariableLengthBytes SizeEncoding
+    deriving (Show)
+
+-- | Tag type used to label branches in the parse tree of a Merkle proof.
+type Tag = String
+
+-- | Identifier type used to identify non-terminals in a Merkle schema.
+type NonTerminal = Int
+
+-- | The body of a rewrite rule from a 'MerkleSchema'.
```
Comment (abizjak) [inline haskell-src/Concordium/MerkleProofs.hs:103]: It's unclear waht "rewrite rule" refers to here. What is the system?

Or in other words, this comment was not illuminating without more context.

**Review by abizjak — COMMENTED** [review]
It is a bit hard to understand the design rationale for this.

It seems quite idiosyncratic, which might be fine, but I would like more rationale about the choices justified on more concrete grounds. Currently everything is very general, but also very specific in some places, such as inclusion of LFMBTree.

There is also very limited testing. Most functionality seems to not be tested, combined with its complexity, this makes it hard to maintain.

I am also surprised we need to be inventing constructs like this for Merkle proofs, which are a standard technology. Perhaps we do.

File: haskell-tests/ConcordiumTests/MerkleProofs.hs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+import Test.Hspec
+
+import qualified Concordium.Crypto.SHA256 as Hash
+import Concordium.MerkleProofs
+
+-- | Block hash of testing block.
+testingBlockHash :: Hash.Hash
+testingBlockHash = read "9926f53cde0d3f25afb2dd9f3eb4050da1b01940c501a3c5d22719535061f95a"
+
+-- | A Merkle proof against the testing block.
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-tests/ConcordiumTests/MerkleProofs.hs:15]: Added some documentation in the file header to address this. Beyond the basic definition of `MerkleProof`, the functionality in `Concordium.MerkleProofs` is largely intended to be used for testing the (future) production of Merkle proofs.

File: haskell-tests/ConcordiumTests/MerkleProofs.hs [hunk 13]
```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Simple unit tests for Merkle proofs.
+--  These tests check that a given Merkle proof hashes and parses to known results.
+--  This is not intended as a comprehensive test of the Merkle proof functionality.
+--  Rather, the proof parsing infrastructure is intended to be used to test the correct
+--  production of Merkle proofs in the node.
```
Comment (abizjak) [inline haskell-tests/ConcordiumTests/MerkleProofs.hs:7]: I don't really see this being done in the node in a comprehensive way.

**Review by abizjak — APPROVED** [review]
I still think the amount of testing here and in the related PR in the node is very inadequate, and the complexity of the introduced types, together with lack of explanatory examples is a problem for maintainability.

There are all these combinators defined, but it's unclear what is even covered in the one test that is there.

I think this suffers from being overly general without explaining what problem it is solving. Concrete examples would help with that.

**Review by EmilLa1 — APPROVED** [review]

File: haskell-src/Concordium/MerkleProofs.hs [hunk 14]
```diff
@@ -0,0 +1,604 @@
...
+merkleBuilder (RawData bytes) = Builder.byteString bytes
+merkleBuilder (SubProof sp) = Builder.shortByteString (SHA256.hashToShortByteString $ toRootHash sp)
+
+-- * Parsing Merkle proofs
+
+-- | How a size of some data or datastructure is encoded in a Merkle proof.
+-- (All sizes are unsigned, since a negative size is meaningless.)
+data SizeEncoding
+    = -- | 16-bit big-endian
+      SizeU16BE
```
[the remark this answers is not in the document]
Comment (td202) [reply haskell-src/Concordium/MerkleProofs.hs:82]: It's not needed, so there was no reason to add it.

**Pull request merged**